Flaky test `TestStoreGateway_InitialSyncWithWaitRingTokensStability`

**Describe the bug**

Test `TestStoreGateway_InitialSyncWithWaitRingTokensStability` is flaky, I've seen it failing multiple times on CI:

- https://github.com/grafana/mimir/runs/4626647705?check_suite_focus=true

```
--- FAIL: TestStoreGateway_InitialSyncWithWaitRingTokensStability (2.07s)
    --- FAIL: TestStoreGateway_InitialSyncWithWaitRingTokensStability/shuffle_sharding_strategy,_20_gateways,_RF_=_3,_SS_=_3_(bucket_index_enabled_=_true) (21.52s)
        gateway_test.go:339: random generator seed: 1640348165248718131
        gateway_test.go:397: 
            	Error Trace:	gateway_test.go:397
            	Error:      	Not equal: 
            	            	expected: 72
            	            	actual  : 84
            	Test:       	TestStoreGateway_InitialSyncWithWaitRingTokensStability/shuffle_sharding_strategy,_20_gateways,_RF_=_3,_SS_=_3_(bucket_index_enabled_=_true)
        gateway_test.go:401: 
            	Error Trace:	gateway_test.go:401
            	Error:      	Not equal: 
            	            	expected: 72
            	            	actual  : 84
            	Test:       	TestStoreGateway_InitialSyncWithWaitRingTokensStability/shuffle_sharding_strategy,_20_gateways,_RF_=_3,_SS_=_3_(bucket_index_enabled_=_true)
        gateway_test.go:402: 
            	Error Trace:	gateway_test.go:402
            	Error:      	Not equal: 
            	            	expected: 6
            	            	actual  : 7
            	Test:       	TestStoreGateway_InitialSyncWithWaitRingTokensStability/shuffle_sharding_strategy,_20_gateways,_RF_=_3,_SS_=_3_(bucket_index_enabled_=_true)
FAIL
```

- https://github.com/grafana/mimir/runs/4627394677?check_suite_focus=true

```
--- FAIL: TestStoreGateway_InitialSyncWithWaitRingTokensStability (2.20s)
    --- FAIL: TestStoreGateway_InitialSyncWithWaitRingTokensStability/shuffle_sharding_strategy,_20_gateways,_RF_=_3,_SS_=_3_(bucket_index_enabled_=_false) (24.75s)
        gateway_test.go:339: random generator seed: 1640356571473108699
        gateway_test.go:402: 
            	Error Trace:	gateway_test.go:402
            	Error:      	Not equal: 
            	            	expected: 6
            	            	actual  : 7
            	Test:       	TestStoreGateway_InitialSyncWithWaitRingTokensStability/shuffle_sharding_strategy,_20_gateways,_RF_=_3,_SS_=_3_(bucket_index_enabled_=_false)
FAIL
```

**To Reproduce**

I tried to reproduce it locally with no luck.

**Expected behavior**

Resulsts should be consistent.

**Environment:**
Our CI

--- a/pkg/storegateway/gateway_test.go
+++ b/pkg/storegateway/gateway_test.go
@@ -362,7 +362,7 @@ func TestStoreGateway_InitialSyncWithWaitRingTokensStability(t *testing.T) {
 					gatewayCfg.ShardingRing.InstanceID = instanceID
 					gatewayCfg.ShardingRing.InstanceAddr = fmt.Sprintf("127.0.0.%d", i)
 					gatewayCfg.ShardingRing.RingCheckPeriod = time.Hour // Do not check the ring topology changes in this test. We want the initial sync only.
-					gatewayCfg.ShardingRing.WaitStabilityMinDuration = 2 * time.Second
+					gatewayCfg.ShardingRing.WaitStabilityMinDuration = 4 * time.Second
 					gatewayCfg.ShardingRing.WaitStabilityMaxDuration = 30 * time.Second
 					gatewayCfg.ShardingEnabled = true
 					gatewayCfg.ShardingStrategy = testData.shardingStrategy

